Container Discovery Management for CEL Workload Exclude (#33284)
DataDog/documentation: content/en/containers/guide/container-discovery-management.md

{{% /tab %}}
{{< /tabs >}}

## Advanced CEL exclusion

In **Agent v7.73+**, you can use the `cel_workload_exclude` configuration option to filter containers from Autodiscovery. This feature allows you to define [Common Expression Langauge][3] rules to target containers to be excluded from telemetry collection.

The following attributes representing the container object are available for your filtering rules:

| Attribute                   | Description                                                             |
|-----------------------------|-------------------------------------------------------------------------|
| `container.name`            | The name of the container.                                              |
| `container.image.reference` | The full reference of the container image (registry, repo, tag/digest). |
| `container.pod.name`        | The name of the pod running the container.                              |
| `container.pod.namespace`   | The Kubernetes namespace of the pod.                                    |
| `container.pod.annotations` | The annotations applied to the pod (key-value map).                     |

### Configuration structure

> The `cel_workload_exclude` configuration option is structured as a list of rule sets joined by logical **ORs**. Each rule set defines the `products` to exclude and the corresponding CEL `rules` to match against containers.

Reviewer: ```suggestion
The `cel_workload_exclude` configuration option is structured as a list of rule sets joined by logical `OR` operators. Each rule set defines the `products` to exclude and the corresponding CEL `rules` to match against containers.
```

The `products` field accepts `metrics`, `logs`, and `global` (exclude container from all listed products).

<div class="alert alert-danger">

> If the configuration contains structural errors or CEL syntax issues, the Agent exits with an error to prevent collecting unintended telemetry which could impact billing.

Reviewer: ```suggestion
If the configuration contains structural errors or CEL syntax issues, the Agent exits with an error to prevent collecting unintended telemetry that could impact billing.
```

</div>

In the example below, metrics and logs are excluded for any container with `nginx` in its name running in the `staging` namespace. Additionally, logs are excluded for any container running the `redis` image, **or** any container within a pod that has the annotation `low_priority: "true"`. The [agent's configuration file][4] can be directly updated as seen by this example.

```yaml
# datadog.yaml
cel_workload_exclude:
- products: [metrics, logs]
  rules:
    containers:
      - container.name.matches("nginx") && container.pod.namespace == "staging"
- products: [logs]
  rules:
    containers:
      - container.image.reference.matches("redis")
      - container.pod.annotations["low_priority"] == "true"
```

> The CEL-backed workload exclusion can also be configured by providing a **JSON**-formatted environment value to `DD_CEL_WORKLOAD_EXCLUDE`.

Reviewer: ```suggestion
The CEL-backed workload exclusion can also be configured by providing a JSON-formatted environment value to `DD_CEL_WORKLOAD_EXCLUDE`.
```
